python/GafferImageUI/CatalogueUI.py

```diff
@@ -103,6 +103,41 @@ def registeredColumns() :
 
 	return __registeredColumns.keys()
 
+#
+# Value Providers
+#
+
+# Returns a column value provider that retrieves the value of a header from the
+# catalogue items image.  If multiple names are provided, the first present in
+# the image headers will be used, allowing a single column to support several
+# source headers depending on the image's origin.
+def imageHeaderValueProvider( nameOrNames, defaultValue = None ) :
+
+	if isinstance( nameOrNames, basestring ) :
+		nameOrNames = [ nameOrNames, ]
+
+	def provider( image, catalogue ) :
+		metadata = catalogue["out"].metadata()
+		for name in nameOrNames :
+			if name in metadata :
+				return metadata[ name ]
+
+		return defaultValue
+
+	return provider
+
+# Returns a column value provider that retrieves a render context value from
+# the catalogue items image.  If multiple names are provided, the first present
+# context entry will be used.
+def contextValueProvider( nameOrNames, defaultValue = None ) :
```

Comment:
Alternatively, perhaps we should follow the example of PathListingWidget and have a proper Column base class with ContextColumn and MetadataColumn derived classes. This would also address the fact that you didn't want to pass columns directly to `registerColumn()` because named tuples weren't very nice to use.

Reply:
I guess I liked de-coupling the fact that the Catalogue uses a `PathLIstingWidget`  internally, given we're talking about a lot of changes there.

Reply:
https://github.com/GafferHQ/gaffer/pull/3688/commits/de91b30ea03e6e24f1b55a52ac5d4d6086dc6b1e, https://github.com/GafferHQ/gaffer/pull/3688/commits/3835bac98d0a9f594273a44f4d0b49ce79d04e15 et. al. refactor to the Class approach.